gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/FlatMapStep.java

```diff
@@ -44,6 +45,7 @@ public FlatMapStep(final Traversal.Admin traversal) {
                 return this.head.split(this.iterator.next(), this);
             } else {
                 this.head = this.starts.next();
+                closeIterator();
```

Comment:
This should go before `this.starts.next()` as if `this.starts.next()` is empty, it will throw `FastNoSuchElementException` and the previous iterator would not have been closed.